dbms/src/Functions/FunctionsString.cpp

```diff
@@ -5782,6 +5781,176 @@ class FunctionElt : public IFunction
+                format<DecimalColVec, IntType>(col, precision, scale, col_res->getChars(), col_res->getOffsets());
+                block.getByPosition(result).column = std::move(col_res);
+                return true;
+            }
+            else
+            {
+                return false;
+            }
+        });
+
+        if (!is_types_valid)
+            throw Exception(
+                fmt::format("Illegal types {} arguments of function {}", base_type->getName(), getName()),
+                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+    }
+
+private:
+    template <typename F>
+    static bool getDecimalType(DataTypePtr type, F && f)
+    {
+        return castTypeToEither<
+            DataTypeDecimal32,
+            DataTypeDecimal64,
+            DataTypeDecimal128,
+            DataTypeDecimal256>(type.get(), std::forward<F>(f));
+    }
+
+    template <typename DecimalColVec, typename IntType>
+    static void format(
+        const DecimalColVec * col,
+        PrecType precision,
+        ScaleType scale,
+        ColumnString::Chars_t & res_data,
+        ColumnString::Offsets & res_offsets)
+    {
+        auto & data = col->getData();
+        size_t size = data.size();
+        res_data.resize(size * (precision + 3));
+        res_offsets.resize(size);
+
+        ColumnString::Offset cur_offset = 0;
+        for (size_t i = 0; i < size; ++i)
+        {
+            const auto & decimal = data[i];
+            if (decimal.value == std::numeric_limits<IntType>::min())
+                // for IntType::min, `value = -value` may cause overflow, so use Int256 here.
+                doFormat<Int256>(decimal.value, scale, cur_offset, res_data, res_offsets[i]);
+            else
+                doFormat<IntType>(decimal.value, scale, cur_offset, res_data, res_offsets[i]);
+        }
+    }
+
+    template <typename IntType>
+    static void doFormat(
+        IntType value,
+        ScaleType scale,
+        ColumnString::Offset & cur_offset,
+        ColumnString::Chars_t & res_data,
+        ColumnString::Offset & res_offset)
+    {
+        if (value < 0)
+        {
+            res_data[cur_offset++] = '-';
+            value = -value;
+        }
+        if (scale > 0 && value > 0)
+        {
+            bool add_dot = false;
+            size_t scale_i = 0;
+            // Ignore trailing zero.
+            while (value > 0 && scale_i < scale)
+            {
+                int d = static_cast<int>(value % 10);
+                value /= 10;
+                ++scale_i;
+                if (d != 0)
+                {
+                    res_data[cur_offset++] = d + '0';
+                    add_dot = true;
+                    break;
+                }
+            }
+            for (; value > 0 && scale_i < scale; ++scale_i)
+            {
+                int d = static_cast<int>(value % 10);
+                value /= 10;
+                res_data[cur_offset++] = d + '0';
+                add_dot = true;
```

Comment:
Looks like there is no need to set `add_dot` to true here.

Reply:
good point!
updated.